File: apps/visualizer/backend/api/identity.py
"""Identity API: best-photos ranking, mirror signature, post-next suggestions (phase 08-01).

Thumbnails are not inlined here; responses include ``image_key`` and ``filename`` so the
frontend can use the existing catalog thumbnail/image routes.
"""

from __future__ import annotations

import sqlite3

from flask import Blueprint, jsonify, request
from flask.typing import ResponseReturnValue
from spectree import Response
from utils.db import with_db
from utils.responses import error_bad_request, error_server_error

from api.openapi import spec
from api.schemas.identity import (
    IdentityBestPhotosQuery,
    IdentityBestPhotosResponse,
    MirrorResponse,
    PostNextSuggestionsQuery,
    PostNextSuggestionsResponse,
)
from api.schemas.jobs import ErrorBody
from utils.pagination import _clamp_pagination
from lightroom_tagger.core.identity_service import (
    build_mirror,
    rank_best_photos,
    suggest_what_to_post_next,
)

bp = Blueprint('identity', __name__)


def _parse_optional_min_perspectives() -> tuple[int | None, ResponseReturnValue | None]:
    raw = request.args.get("min_perspectives")
    if raw is None or str(raw).strip() == "":
        return None, None
    try:
        v = int(str(raw).strip())
    except (TypeError, ValueError):
        return None, error_bad_request("min_perspectives must be an integer")
    if v < 1:
        return None, error_bad_request("min_perspectives must be at least 1")
    if v > 50:
        v = 50
    return v, None


def _parse_sort_by_date() -> tuple[str | None, ResponseReturnValue | None]:
    raw = (request.args.get("sort_by_date") or "").strip().lower()
    if not raw:
        return None, None
    if raw not in ("newest", "oldest"):
        return None, error_bad_request("sort_by_date must be newest or oldest")
    return raw, None


def _parse_optional_posted() -> tuple[bool | None, ResponseReturnValue | None]:
    raw = request.args.get("posted")
    if raw is None or str(raw).strip() == "":
        return None, None
    key = str(raw).strip().lower()
    if key in ("true", "1", "yes"):
        return True, None
    if key in ("false", "0", "no"):
        return False, None
    return None, error_bad_request("posted must be true or false")


@bp.route("/best-photos", methods=["GET"])
@with_db
@spec.validate(
    query=IdentityBestPhotosQuery,
    resp=Response(HTTP_200=IdentityBestPhotosResponse, HTTP_400=ErrorBody, HTTP_500=ErrorBody),
    tags=['identity'],
)
def best_photos(db: sqlite3.Connection):
    """Paginated eligible catalog images ranked by peak within-perspective percentile."""
    try:
        limit, offset = _clamp_pagination(
            request.args.get("limit", 50, type=int),
            request.args.get("offset", 0, type=int),
        )
        min_p, err = _parse_optional_min_perspectives()
        if err is not None:
            return err
        sort_by_date, err2 = _parse_sort_by_date()
        if err2 is not None:
            return err2
        posted_value, err3 = _parse_optional_posted()
        if err3 is not None:
            return err3

        items, total, meta = rank_best_photos(
            db,
            limit=limit,
            offset=offset,
            min_perspectives=min_p,
            sort_by_date=sort_by_date,
            posted=posted_value,
        )
        return jsonify({"items": items, "total": total, "meta": meta})
    except Exception:
        return error_server_error()


@bp.route("/mirror", methods=["GET"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=MirrorResponse, HTTP_500=ErrorBody),
    tags=['identity'],
)
def mirror(db: sqlite3.Connection):
    """Catalog Mirror: crowned signature techniques and exemplar rails."""
    try:
        return jsonify(build_mirror(db))
    except Exception:
        return error_server_error()


@bp.route("/suggestions", methods=["GET"])
@with_db
@spec.validate(
    query=PostNextSuggestionsQuery,
    resp=Response(HTTP_200=PostNextSuggestionsResponse, HTTP_400=ErrorBody, HTTP_500=ErrorBody),
    tags=['identity'],
)
def suggestions(db: sqlite3.Connection):
    """What to post next: unposted, coverage-eligible images with reason codes."""
    try:
        limit, offset = _clamp_pagination(
            request.args.get("limit", 20, type=int),
            request.args.get("offset", 0, type=int),
        )
        sort_by_date, err = _parse_sort_by_date()
        if err is not None:
            return err

        payload = suggest_what_to_post_next(
            db,
            limit=limit,
            offset=offset,
            sort_by_date=sort_by_date,
        )
        return jsonify(
            {
                "candidates": payload["candidates"],
                "total": payload["total"],
                "meta": payload["meta"],
                "empty_state": payload["empty_state"],
            }
        )
    except Exception:
        return error_server_error()

File: apps/visualizer/backend/src/api/identity.ts
/**
 * Identity API: best-photos ranking, mirror signature, post-next suggestions.
 *
 * Thumbnails are not inlined: responses carry `image_key` and `filename` so the
 * frontend reuses the catalog thumbnail route.
 *
 * Query parameters on these routes appear in the OpenAPI contract. Integer params
 * use `z.coerce` so string query values like `"4"` parse as integers.
 */
import { createRoute, z } from '@hono/zod-openapi';
import { libraryDb, type LibraryEnv } from './library-db.js';
import { buildLensExemplars, buildMirror } from '../identity/mirror.js';
import { rankBestPhotos } from '../identity/ranking.js';
import { suggestWhatToPostNext } from '../identity/suggest-post.js';
import { clampPagination } from '../utils/responses.js';
import { createOpenApiApp } from './openapi.js';
import { jsonBody, withValidationError } from './route-helpers.js';
import { ErrorBody } from './schemas/errors.js';
import {
  IdentityBestPhotosResponse,
  MirrorLensExemplarsResponse,
  MirrorResponse,
  PostNextSuggestionsResponse,
} from './schemas/identity.js';

export const identityRoutes = createOpenApiApp<LibraryEnv>();

/** An optional integer query parameter, emitted as `integer | null`. */
const intParam = z.coerce.number().int().nullish();

identityRoutes.use('/identity/*', libraryDb());

type Parsed<T> = { ok: true; value: T } | { ok: false; error: string };

/**
 * `min_perspectives`: at least 1, silently capped at 50.
 *
 * `intParam` has already rejected a non-integer with 422, so the floor and the cap
 * are all that is left. The cap is a clamp rather than a rejection — asking for 200
 * perspectives is not a client error, there just are not that many.
 */
function parseMinPerspectives(raw: number | null | undefined): Parsed<number | null> {
  if (raw === null || raw === undefined) return { ok: true, value: null };
  if (raw < 1) return { ok: false, error: 'min_perspectives must be at least 1' };
  return { ok: true, value: Math.min(raw, 50) };
}

function parseSortByDate(raw: string | null | undefined): Parsed<'newest' | 'oldest' | null> {
  const value = (raw ?? '').trim().toLowerCase();
  if (!value) return { ok: true, value: null };
  if (value !== 'newest' && value !== 'oldest') {
    return { ok: false, error: 'sort_by_date must be newest or oldest' };
  }
  return { ok: true, value };
}

/**
 * `posted`: a tri-state.
 *
 * Unlike the catalog list — which silently ignores an unrecognized value — this one
 * rejects it, and accepts `1`/`yes`/`0`/`no` as well as the booleans.
 */
function parsePosted(raw: string | null | undefined): Parsed<boolean | null> {
  if (raw === null || raw === undefined || raw.trim() === '') return { ok: true, value: null };
  const key = raw.trim().toLowerCase();
  if (['true', '1', 'yes'].includes(key)) return { ok: true, value: true };
  if (['false', '0', 'no'].includes(key)) return { ok: true, value: false };
  return { ok: false, error: 'posted must be true or false' };
}

// --- best photos ------------------------------------------------------------

const bestPhotosRoute = createRoute({
  method: 'get',
  path: '/identity/best-photos',
  tags: ['identity'],
  request: {
    query: z.object({
      limit: intParam,
      offset: intParam,
      min_perspectives: intParam,
      sort_by_date: z.string().nullish(),
      posted: z.string().nullish(),
    }),
  },
  responses: withValidationError({
    200: {
      description: 'Ranked eligible catalog images',
      content: jsonBody(IdentityBestPhotosResponse),
    },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

identityRoutes.openapi(bestPhotosRoute, (c) => {
  const query = c.req.valid('query');
  const { limit, offset } = clampPagination(query.limit, query.offset);
  const minP = parseMinPerspectives(query.min_perspectives);
  if (!minP.ok) return c.json({ error: minP.error }, 400);
  const sortByDate = parseSortByDate(query.sort_by_date);
  if (!sortByDate.ok) return c.json({ error: sortByDate.error }, 400);
  const posted = parsePosted(query.posted);
  if (!posted.ok) return c.json({ error: posted.error }, 400);

  const { items, total, meta } = rankBestPhotos(c.get('libraryDb'), {
    limit,
    offset,
    minPerspectives: minP.value,
    sortByDate: sortByDate.value,
    posted: posted.value,
  });
  return c.json({ items, total, meta } as z.infer<typeof IdentityBestPhotosResponse>, 200);
});

// --- mirror -----------------------------------------------------------------

const mirrorRoute = createRoute({
  method: 'get',
  path: '/identity/mirror',
  tags: ['identity'],
  responses: withValidationError({
    200: { description: 'Crowned signature techniques', content: jsonBody(MirrorResponse) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

identityRoutes.openapi(mirrorRoute, (c) =>
  c.json(buildMirror(c.get('libraryDb')) as z.infer<typeof MirrorResponse>, 200),
);

const lensExemplarsRoute = createRoute({
  method: 'get',
  path: '/identity/mirror/lens/{slug}/exemplars',
  tags: ['identity'],
  request: {
    params: z.object({ slug: z.string() }),
    query: z.object({ limit: intParam, offset: intParam }),
  },
  responses: withValidationError({
    200: { description: 'Exemplar rail', content: jsonBody(MirrorLensExemplarsResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

identityRoutes.openapi(lensExemplarsRoute, (c) => {
  const query = c.req.valid('query');
  // Default 24, not the shared helper's 50 — it is the initial exemplar rail size.
  const { limit, offset } = clampPagination(query.limit, query.offset, 24);
  try {
    return c.json(
      buildLensExemplars(c.get('libraryDb'), c.req.valid('param').slug, { offset, limit }),
      200,
    );
  } catch (e) {
    // An unknown or deactivated slug is a 400.
    if (e instanceof RangeError) return c.json({ error: e.message }, 400);
    throw e;
  }
});

// --- suggestions ------------------------------------------------------------

const suggestionsRoute = createRoute({
  method: 'get',
  path: '/identity/suggestions',
  tags: ['identity'],
  request: {
    query: z.object({
      limit: intParam,
      offset: intParam,
      sort_by_date: z.string().nullish(),
    }),
  },
  responses: withValidationError({
    200: {
      description: 'What to post next',
      content: jsonBody(PostNextSuggestionsResponse),
    },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

identityRoutes.openapi(suggestionsRoute, (c) => {
  const query = c.req.valid('query');
  // Default 20 here.
  const { limit, offset } = clampPagination(query.limit, query.offset, 20);
  const sortByDate = parseSortByDate(query.sort_by_date);
  if (!sortByDate.ok) return c.json({ error: sortByDate.error }, 400);

  const payload = suggestWhatToPostNext(c.get('libraryDb'), {
    limit,
    offset,
    sortByDate: sortByDate.value,
  });
  return c.json(
    {
      candidates: payload.candidates,
      total: payload.total,
      meta: payload.meta,
      empty_state: payload.empty_state,
    } as z.infer<typeof PostNextSuggestionsResponse>,
    200,
  );
});
